pytrip/res/concave_tool_v2.py

```diff
@@ -0,0 +1,111 @@
+"""
+This code is strongly inspired by:
+https://github.com/ellieb/EGSnrc/blob/DICOM-viewer/HEN_HOUSE/gui/dose-viewer/volumes/structure-set-volume.js
+written by Elise Badun:
+https://github.com/ellieb
+"""
+import numpy as np
+import matplotlib.pyplot as plt
+from pytrip import _cntr
+
+
+def to_indices(mm, pixel_size, offsets, slice_thickness):
+    x_off, y_off, z_off = offsets
+    return [
+        round((mm[0] - x_off) / pixel_size),
+        round((mm[1] - y_off) / pixel_size),
+        round((mm[2] - z_off) / slice_thickness)
+    ]
+
+
+def initialize_bitmap(plane, cube_size):
+    x_size, y_size, z_size = cube_size
+    bitmap = None
+    if plane == 'Sagittal':
+        bitmap = np.zeros((z_size, y_size))
+    elif plane == 'Coronal':
+        bitmap = np.zeros((z_size, x_size))
+    return bitmap
+
+
+def fill_bitmap(bitmap, intersections_list_filtered, offsets, pixel_size, plane, slice_thickness):
+    for intersection in intersections_list_filtered:
+        for i in range(1, len(intersection), 2):
+            point_0 = to_indices(intersection[i - 1], pixel_size, offsets, slice_thickness)
+            point_1 = to_indices(intersection[i], pixel_size, offsets, slice_thickness)
+            y = point_0[2]
+            x_0, x_1 = None, None
+            if plane == 'Sagittal':
+                x_0 = point_0[1]
+                x_1 = point_1[1]
+            elif plane == 'Coronal':
+                x_0 = point_0[0]
+                x_1 = point_1[0]
+            bitmap[y, x_0:x_1] = 1
+
+
+def get_depth(intersections_list_filtered, plane):
+    if plane == 'Sagittal':
+        return intersections_list_filtered[0][0][0]
+    if plane == 'Coronal':
+        return intersections_list_filtered[0][0][1]
+
+
+def translate_contour_to_mm(contours_indices, depth, offsets, pixel_size, plane, slice_thickness):
+    contours_mm = []
+    x_offset, y_offset, z_offset = offsets
+    for v in contours_indices:
+        y = v[:, 1] * slice_thickness + z_offset
+        contour_mm = None
+        if plane == 'Sagittal':
+            x = v[:, 0] * pixel_size + y_offset
+            zipped_xy = zip(x, y)
+            contour_mm = [[depth, real_y, real_z] for real_y, real_z in zipped_xy]
+        elif plane == 'Coronal':
+            x = v[:, 0] * pixel_size + x_offset
+            zipped_xy = zip(x, y)
+            contour_mm = [[real_x, depth, real_z] for real_x, real_z in zipped_xy]
+        if contour_mm:
+            contours_mm.append(contour_mm)
+    return contours_mm
+
+
+def calculate_contour(bitmap):
+    fig, ax = plt.subplots(1, 1)
+    cp = ax.contour(bitmap, levels=0)
+    paths = cp.collections[0].get_paths()
+    contours = [np.array(p.vertices) for p in paths]
+
+    return contours
+
+
+def calculate_contour_v2(bitmap):

```

Comment:
How many times this method is being called, i.e. for `External` contour ?

Reply:
I'd also reccommend profiling this method line-line for worst case situation (i.e. Extrernal contour).
Some parts may be not optimal, like creation of `meshgrid` - does bitmap have the same shape for every evaluation ?
Also final for loop may be expensive

Reply:
one call per slice where contour should exist

Reply:
also:
![image](https://user-images.githubusercontent.com/44035398/147251720-9c5362e9-918c-43ae-a669-ae29abfbf037.png)
contouring does not take as long as intersecting does

Reply:
which method is being called to calculate intersecting ?

Reply:
![image](https://user-images.githubusercontent.com/44035398/147257489-14ce771d-abdd-4d4c-9dc9-a7f86d1f0195.png)
`create_intersections' i think it's slow because of number of segments it has to transverse in order to find intersection

Reply:
Can you try to speed it up locally using numba (https://numba.pydata.org/) ?
This way we can estimate if it is worth to be translated to C

Reply:
we fixed C code